pkg/client/event_deduplicator.go

```diff
@@ -0,0 +1,188 @@
+package client
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	chain "github.com/keep-network/keep-ecdsa/pkg/chain"
+)
+
+// eventDeduplicator decides whether the given event should be handled by the
+// client or not.
+//
+// Event subscription may emit the same event two or more times. The same event
+// can be emitted right after it's been emitted for the first time. The same
+// event can also be emitted a long time after it's been emitted for the first
+// time. It is eventDeduplicator's responsibility to decide whether the given
+// event is a duplicate and should be ignored or if it is not a duplicate and
+// should be handled.
+//
+// Four events are supported:
+// - key generation request for a new keep,
+// - signature request for a keep,
+// - keep close request,
+// - keep terminate request.
+type eventDeduplicator struct {
```

Comment:
If we decide to use `string` instead of `common.Address` in the event tracker, we should also do it here. I see we will have a problem with the `keepRegistry` interface but we can replace it by a type `type signerExistsFn func(keepAddress string) bool` or something similiar and let the caller play with address conversion.

Reply:
Yesh... I was struggling with it, took another attempt and... I am not sure if it's the right moment to address it. 

`keepRegistry` is not the only problem. The entire `chain.Handle` interface is organized around `common.Address`. So it is `SubmitSignature`, `SubmitKeepPublicKey`, also `RequestSignature`, `AddSigner` and `OpenKeep` from `local` chain implementation.

As much as I do agree we should not let the `common.Address` propagate any further and try to write a new code in a chain-agnostic way I do wonder if this situation is not an exception. We'll end up with a bunch of transformations back and forward.

Instead of doing it now, I'd suggest addressing `common.Address` problem separately in another PR while preparing the client for proper multi-chain support where we'll update all the layers. What do you think?

Reply:
Actually, I think we can try to address it partially. Leave `keepRegistry` with `common.Address` and `common.HexToAddress` before calling it. The rest of the code are tests, so there might be some acceptable level of boilerplate code to achieve our goal. Let me try, I am not satisfied with what's here currently.

Reply:
Sure that's a reasonable proposition.

Reply:
Here is the proposition, I intentionally pushed it to a separate branch in case we decide not to go with it here: https://github.com/keep-network/keep-ecdsa/commit/bdcfb776b35216ffb6cfc78516916657bdd2c399

I am not sure... there is a lot of going back and forward with transformation. I feel that the first comment I made in this thread makes sense and maybe we should wait with this change and swallow `common.Address` for now. There are too many dependencies with other components in the deduplicator.

Reply:
Ok, after looking at the proposition I agree we should rather leave `common.Address` for now. It's safer and more readable for now. Trying to remove that dependency doesn't give us any benefits for now. We will have to revisit that part while implementing chain-agnostic approach anyway.